File: backend/src/models/CloudSync.js
const mongoose = require('mongoose');

// Main cloud sync record for tracking user data synchronization
const CloudSyncSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    index: true
  },
  
  // Sync metadata
  lastSyncAt: {
    type: Date,
    default: Date.now
  },
  nextSyncAt: {
    type: Date,
    default: () => new Date(Date.now() + 3600000) // 1 hour from now
  },
  syncStatus: {
    type: String,
    enum: ['idle', 'syncing', 'error', 'conflict', 'paused'],
    default: 'idle'
  },
  syncVersion: {
    type: Number,
    default: 1
  },
  
  // Device tracking
  devices: [{
    deviceId: {
      type: String,
      required: true
    },
    deviceName: String,
    platform: {
      type: String,
      enum: ['ios', 'android', 'web'],
      required: true
    },
    lastActiveAt: {
      type: Date,
      default: Date.now
    },
    syncEnabled: {
      type: Boolean,
      default: true
    }
  }],
  
  // Data sync status for different types
  dataSyncStatus: {
    photos: {
      lastSyncAt: Date,
      totalItems: { type: Number, default: 0 },
      syncedItems: { type: Number, default: 0 },
      pendingItems: { type: Number, default: 0 },
      failedItems: { type: Number, default: 0 }
    },
    presets: {
      lastSyncAt: Date,
      totalItems: { type: Number, default: 0 },
      syncedItems: { type: Number, default: 0 },
      pendingItems: { type: Number, default: 0 },
      failedItems: { type: Number, default: 0 }
    },
    settings: {
      lastSyncAt: Date,
      totalItems: { type: Number, default: 0 },
      syncedItems: { type: Number, default: 0 },
      pendingItems: { type: Number, default: 0 },
      failedItems: { type: Number, default: 0 }
    }
  },
  
  // Storage usage
  storageUsage: {
    totalBytes: { type: Number, default: 0 },
    photosBytes: { type: Number, default: 0 },
    thumbnailsBytes: { type: Number, default: 0 },
    quotaBytes: { type: Number, default: 5368709120 }, // 5GB default
    lastCalculatedAt: Date
  },
  
  // Sync settings
  syncSettings: {
    autoSync: { type: Boolean, default: true },
    syncOnWiFiOnly: { type: Boolean, default: true },
    syncPhotos: { type: Boolean, default: true },
    syncPresets: { type: Boolean, default: true },
    syncSettings: { type: Boolean, default: true },
    compressionLevel: {
      type: String,
      enum: ['none', 'low', 'medium', 'high'],
      default: 'medium'
    },
    syncFrequency: {
      type: String,
      enum: ['realtime', 'hourly', 'daily', 'manual'],
      default: 'hourly'
    }
  },
  
  // Error tracking
  lastError: {
    message: String,
    code: String,
    timestamp: Date,
    retryCount: { type: Number, default: 0 }
  },
  
  // Conflict resolution
  conflicts: [{
    type: {
      type: String,
      enum: ['photo', 'preset', 'setting'],
      required: true
    },
    localId: String,
    cloudId: String,
    conflictReason: String,
    localVersion: mongoose.Schema.Types.Mixed,
    cloudVersion: mongoose.Schema.Types.Mixed,
    detectedAt: { type: Date, default: Date.now },
    resolvedAt: Date,
    resolution: {
      type: String,
      enum: ['local_wins', 'cloud_wins', 'merge', 'manual']
    }
  }]
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Photo cloud storage record
const CloudPhotoSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    index: true
  },
  
  // Photo identification
  localId: String, // ID from mobile device
  filename: {
    type: String,
    required: true
  },
  originalFilename: String,
  
  // S3 storage info
  s3Key: {
    type: String,
    required: true,
    unique: true
  },
  s3Bucket: String,
  s3Region: String,
  cloudUrl: String,
  thumbnailS3Key: String,
  thumbnailUrl: String,
  
  // Photo metadata
  metadata: {
    fileSize: Number,
    dimensions: {
      width: Number,
      height: Number
    },
    format: String,
    exifData: {
      camera: String,
      lens: String,
      iso: Number,
      aperture: String,
      shutterSpeed: String,
      focalLength: Number,
      capturedAt: Date,
      gpsLocation: {
        latitude: Number,
        longitude: Number
      }
    },
    processingInfo: {
      aiAnalysis: mongoose.Schema.Types.Mixed,
      appliedPreset: String,
      adjustments: mongoose.Schema.Types.Mixed
    }
  },
  
  // Sync status
  syncStatus: {
    type: String,
    enum: ['pending', 'uploading', 'synced', 'failed', 'deleted'],
    default: 'pending'
  },
  uploadProgress: {
    type: Number,
    default: 0,
    min: 0,
    max: 100
  },
  
  // Sharing and access
  isPublic: { type: Boolean, default: false },
  sharedWith: [String], // User IDs
  tags: [String],
  
  // Timestamps
  capturedAt: Date,
  uploadedAt: Date,
  lastModifiedAt: Date
}, {
  timestamps: true,
  index: { userId: 1, syncStatus: 1 }
});

// Cloud settings sync record
const CloudSettingsSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    index: true
  },
  
  // Settings data
  settingsType: {
    type: String,
    enum: ['camera_defaults', 'user_preferences', 'app_settings'],
    required: true
  },
  settingsData: {
    type: mongoose.Schema.Types.Mixed,
    required: true
  },
  
  // Versioning
  version: {
    type: Number,
    default: 1
  },
  checksum: String, // For integrity verification
  
  // Device sync tracking
  deviceSyncStatus: [{
    deviceId: String,
    lastSyncedVersion: Number,
    syncStatus: {
      type: String,
      enum: ['synced', 'pending', 'conflict', 'error'],
      default: 'pending'
    },
    lastSyncAt: Date
  }],
  
  // Sync metadata
  lastModifiedBy: String, // Device ID
  syncPriority: {
    type: String,
    enum: ['low', 'normal', 'high'],
    default: 'normal'
  }
}, {
  timestamps: true,
  index: { userId: 1, settingsType: 1 }
});

// Sync job queue for background processing
const SyncJobSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    index: true
  },
  
  jobType: {
    type: String,
    enum: ['photo_upload', 'photo_download', 'settings_sync', 'preset_sync', 'full_sync'],
    required: true
  },
  
  // Job data
  jobData: {
    type: mongoose.Schema.Types.Mixed,
    required: true
  },
  
  // Job status
  status: {
    type: String,
    enum: ['pending', 'processing', 'completed', 'failed', 'cancelled'],
    default: 'pending'
  },
  priority: {
    type: Number,
    default: 5,
    min: 1,
    max: 10
  },
  
  // Execution info
  attempts: { type: Number, default: 0 },
  maxAttempts: { type: Number, default: 3 },
  processingStartedAt: Date,
  completedAt: Date,
  
  // Progress tracking
  progress: {
    current: { type: Number, default: 0 },
    total: { type: Number, default: 100 },
    message: String
  },
  
  // Error handling
  error: {
    message: String,
    stack: String,
    code: String
  },
  
  // Scheduling
  scheduledFor: {
    type: Date,
    default: Date.now
  },
  expiresAt: {
    type: Date,
    default: () => new Date(Date.now() + 24 * 60 * 60 * 1000) // 24 hours
  }
}, {
  timestamps: true,
  index: { status: 1, scheduledFor: 1, priority: -1 }
});

// Add virtual properties and methods
CloudSyncSchema.virtual('totalSyncedItems').get(function() {
  const { photos, presets, settings } = this.dataSyncStatus;
  return (photos.syncedItems || 0) + (presets.syncedItems || 0) + (settings.syncedItems || 0);
});

CloudSyncSchema.virtual('totalPendingItems').get(function() {
  const { photos, presets, settings } = this.dataSyncStatus;
  return (photos.pendingItems || 0) + (presets.pendingItems || 0) + (settings.pendingItems || 0);
});

CloudSyncSchema.virtual('syncProgress').get(function() {
  const total = this.totalSyncedItems + this.totalPendingItems;
  return total > 0 ? (this.totalSyncedItems / total) * 100 : 100;
});

CloudSyncSchema.virtual('storageUsagePercent').get(function() {
  const { totalBytes, quotaBytes } = this.storageUsage;
  return quotaBytes > 0 ? (totalBytes / quotaBytes) * 100 : 0;
});

// Methods
CloudSyncSchema.methods.addDevice = function(deviceInfo) {
  const existingDevice = this.devices.find(d => d.deviceId === deviceInfo.deviceId);
  if (existingDevice) {
    Object.assign(existingDevice, deviceInfo, { lastActiveAt: new Date() });
  } else {
    this.devices.push({ ...deviceInfo, lastActiveAt: new Date() });
  }
  return this.save();
};

CloudSyncSchema.methods.updateSyncStatus = function(dataType, status) {
  if (!this.dataSyncStatus[dataType]) {
    this.dataSyncStatus[dataType] = {};
  }
  Object.assign(this.dataSyncStatus[dataType], status, { lastSyncAt: new Date() });
  return this.save();
};

CloudSyncSchema.methods.addConflict = function(conflictData) {
  this.conflicts.push(conflictData);
  this.syncStatus = 'conflict';
  return this.save();
};

CloudPhotoSchema.methods.updateUploadProgress = function(progress) {
  this.uploadProgress = Math.min(100, Math.max(0, progress));
  if (progress >= 100) {
    this.syncStatus = 'synced';
    this.uploadedAt = new Date();
  }
  return this.save();
};

// Static methods
CloudSyncSchema.statics.findByUserId = function(userId) {
  return this.findOne({ userId }).populate('userId', 'name email');
};

CloudPhotoSchema.statics.findPendingUploads = function(userId, limit = 10) {
  return this.find({ 
    userId, 
    syncStatus: { $in: ['pending', 'uploading'] }
  }).limit(limit);
};

SyncJobSchema.statics.getNextJob = function() {
  return this.findOneAndUpdate(
    { 
      status: 'pending',
      scheduledFor: { $lte: new Date() },
      attempts: { $lt: this.$where('this.maxAttempts') }
    },
    { 
      status: 'processing',
      processingStartedAt: new Date(),
      $inc: { attempts: 1 }
    },
    { 
      new: true,
      sort: { priority: -1, scheduledFor: 1 }
    }
  );
};

// Indexes for performance
CloudSyncSchema.index({ userId: 1 }, { unique: true });
CloudSyncSchema.index({ 'devices.deviceId': 1 });
CloudSyncSchema.index({ syncStatus: 1, nextSyncAt: 1 });

CloudPhotoSchema.index({ userId: 1, syncStatus: 1 });
CloudPhotoSchema.index({ s3Key: 1 }, { unique: true });
CloudPhotoSchema.index({ capturedAt: -1 });

CloudSettingsSchema.index({ userId: 1, settingsType: 1 }, { unique: true });
CloudSettingsSchema.index({ version: -1 });

SyncJobSchema.index({ userId: 1, status: 1 });
SyncJobSchema.index({ status: 1, scheduledFor: 1, priority: -1 });
SyncJobSchema.index({ expiresAt: 1 }, { expireAfterSeconds: 0 });

module.exports = {
  CloudSync: mongoose.model('CloudSync', CloudSyncSchema),
  CloudPhoto: mongoose.model('CloudPhoto', CloudPhotoSchema),
  CloudSettings: mongoose.model('CloudSettings', CloudSettingsSchema),
  SyncJob: mongoose.model('SyncJob', SyncJobSchema)
};

File: mobile/lib/models/cloud_sync.dart

import 'dart:math' show pow;

class CloudSyncStatus {
  final String syncStatus;
  final DateTime? lastSyncAt;
  final DateTime? nextSyncAt;
  final double syncProgress;
  final DataSyncStatus dataSyncStatus;
  final StorageUsage storageUsage;
  final List<SyncDevice> devices;
  final List<SyncConflict> conflicts;
  final SyncSettings syncSettings;

  CloudSyncStatus({
    required this.syncStatus,
    this.lastSyncAt,
    this.nextSyncAt,
    required this.syncProgress,
    required this.dataSyncStatus,
    required this.storageUsage,
    required this.devices,
    required this.conflicts,
    required this.syncSettings,
  });

  factory CloudSyncStatus.fromJson(Map<String, dynamic> json) {
    return CloudSyncStatus(
      syncStatus: json['syncStatus'] ?? 'idle',
      lastSyncAt: json['lastSyncAt'] != null ? DateTime.parse(json['lastSyncAt']) : null,
      nextSyncAt: json['nextSyncAt'] != null ? DateTime.parse(json['nextSyncAt']) : null,
      syncProgress: (json['syncProgress'] ?? 0.0).toDouble(),
      dataSyncStatus: DataSyncStatus.fromJson(json['dataSyncStatus'] ?? {}),
      storageUsage: StorageUsage.fromJson(json['storageUsage'] ?? {}),
      devices: (json['devices'] as List? ?? [])
          .map((device) => SyncDevice.fromJson(device))
          .toList(),
      conflicts: (json['conflicts'] as List? ?? [])
          .map((conflict) => SyncConflict.fromJson(conflict))
          .toList(),
      syncSettings: SyncSettings.fromJson(json['syncSettings'] ?? {}),
    );
  }

  bool get hasConflicts => conflicts.isNotEmpty;
  bool get isSyncing => syncStatus == 'syncing';
  bool get hasErrors => syncStatus == 'error';
}

class DataSyncStatus {
  final SyncItemStatus photos;
  final SyncItemStatus presets;
  final SyncItemStatus settings;

  DataSyncStatus({
    required this.photos,
    required this.presets,
    required this.settings,
  });

  factory DataSyncStatus.fromJson(Map<String, dynamic> json) {
    return DataSyncStatus(
      photos: SyncItemStatus.fromJson(json['photos'] ?? {}),
      presets: SyncItemStatus.fromJson(json['presets'] ?? {}),
      settings: SyncItemStatus.fromJson(json['settings'] ?? {}),
    );
  }

  int get totalSyncedItems => photos.syncedItems + presets.syncedItems + settings.syncedItems;
  int get totalPendingItems => photos.pendingItems + presets.pendingItems + settings.pendingItems;
  int get totalFailedItems => photos.failedItems + presets.failedItems + settings.failedItems;
}

class SyncItemStatus {
  final DateTime? lastSyncAt;
  final int totalItems;
  final int syncedItems;
  final int pendingItems;
  final int failedItems;

  SyncItemStatus({
    this.lastSyncAt,
    required this.totalItems,
    required this.syncedItems,
    required this.pendingItems,
    required this.failedItems,
  });

  factory SyncItemStatus.fromJson(Map<String, dynamic> json) {
    return SyncItemStatus(
      lastSyncAt: json['lastSyncAt'] != null ? DateTime.parse(json['lastSyncAt']) : null,
      totalItems: json['totalItems'] ?? 0,
      syncedItems: json['syncedItems'] ?? 0,
      pendingItems: json['pendingItems'] ?? 0,
      failedItems: json['failedItems'] ?? 0,
    );
  }

  double get progress {
    if (totalItems == 0) return 1.0;
    return syncedItems / totalItems;
  }
}

class StorageUsage {
  final int totalBytes;
  final int photosBytes;
  final int thumbnailsBytes;
  final int quotaBytes;
  final DateTime? lastCalculatedAt;
  final int objectCount;
  final String formattedSize;

  StorageUsage({
    required this.totalBytes,
    required this.photosBytes,
    required this.thumbnailsBytes,
    required this.quotaBytes,
    this.lastCalculatedAt,
    required this.objectCount,
    required this.formattedSize,
  });

  factory StorageUsage.fromJson(Map<String, dynamic> json) {
    return StorageUsage(
      totalBytes: json['totalBytes'] ?? 0,
      photosBytes: json['photosBytes'] ?? 0,
      thumbnailsBytes: json['thumbnailsBytes'] ?? 0,
      quotaBytes: json['quotaBytes'] ?? 5368709120, // 5GB default
      lastCalculatedAt: json['lastCalculatedAt'] != null 
          ? DateTime.parse(json['lastCalculatedAt']) 
          : null,
      objectCount: json['objectCount'] ?? 0,
      formattedSize: json['formattedSize'] ?? '0 Bytes',
    );
  }

  double get usagePercent {
    if (quotaBytes == 0) return 0.0;
    return (totalBytes / quotaBytes).clamp(0.0, 1.0);
  }

  String get quotaFormatted => _formatBytes(quotaBytes);
  
  bool get isNearQuota => usagePercent > 0.8;
  bool get isOverQuota => usagePercent >= 1.0;

  static String _formatBytes(int bytes) {
    if (bytes == 0) return '0 Bytes';
    
    const k = 1024;
    const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB'];
    final i = (bytes.bitLength - 1) ~/ 10;
    
    return '${(bytes / pow(k, i)).toStringAsFixed(2)} ${sizes[i]}';
  }
}

class SyncDevice {
  final String deviceId;
  final String? deviceName;
  final String platform;
  final DateTime lastActiveAt;
  final bool syncEnabled;

  SyncDevice({
    required this.deviceId,
    this.deviceName,
    required this.platform,
    required this.lastActiveAt,
    required this.syncEnabled,
  });

  factory SyncDevice.fromJson(Map<String, dynamic> json) {
    return SyncDevice(
      deviceId: json['deviceId'] ?? '',
      deviceName: json['deviceName'],
      platform: json['platform'] ?? 'unknown',
      lastActiveAt: DateTime.parse(json['lastActiveAt'] ?? DateTime.now().toIso8601String()),
      syncEnabled: json['syncEnabled'] ?? true,
    );
  }

  String get displayName => deviceName ?? '${platform.capitalize()} Device';
  bool get isCurrentDevice => deviceId == getCurrentDeviceId();
  
  static String getCurrentDeviceId() {
    // Implementation would return actual device ID
    return 'current_device_id';
  }
}

class SyncConflict {
  final String id;
  final String type;
  final String? localId;
  final String? cloudId;
  final String conflictReason;
  final Map<String, dynamic>? localVersion;
  final Map<String, dynamic>? cloudVersion;
  final DateTime detectedAt;
  final DateTime? resolvedAt;
  final String? resolution;

  SyncConflict({
    required this.id,
    required this.type,
    this.localId,
    this.cloudId,
    required this.conflictReason,
    this.localVersion,
    this.cloudVersion,
    required this.detectedAt,
    this.resolvedAt,
    this.resolution,
  });

  factory SyncConflict.fromJson(Map<String, dynamic> json) {
    return SyncConflict(
      id: json['_id'] ?? json['id'] ?? '',
      type: json['type'] ?? 'unknown',
      localId: json['localId'],
      cloudId: json['cloudId'],
      conflictReason: json['conflictReason'] ?? '',
      localVersion: json['localVersion'],
      cloudVersion: json['cloudVersion'],
      detectedAt: DateTime.parse(json['detectedAt'] ?? DateTime.now().toIso8601String()),
      resolvedAt: json['resolvedAt'] != null ? DateTime.parse(json['resolvedAt']) : null,
      resolution: json['resolution'],
    );
  }

  bool get isResolved => resolvedAt != null;
  String get displayType => type.capitalize();
}

class SyncSettings {
  final bool autoSync;
  final bool syncOnWiFiOnly;
  final bool syncPhotos;
  final bool syncPresets;
  final bool syncSettings;
  final String compressionLevel;
  final String syncFrequency;

  SyncSettings({
    required this.autoSync,
    required this.syncOnWiFiOnly,
    required this.syncPhotos,
    required this.syncPresets,
    required this.syncSettings,
    required this.compressionLevel,
    required this.syncFrequency,
  });

  factory SyncSettings.fromJson(Map<String, dynamic> json) {
    return SyncSettings(
      autoSync: json['autoSync'] ?? true,
      syncOnWiFiOnly: json['syncOnWiFiOnly'] ?? true,
      syncPhotos: json['syncPhotos'] ?? true,
      syncPresets: json['syncPresets'] ?? true,
      syncSettings: json['syncSettings'] ?? true,
      compressionLevel: json['compressionLevel'] ?? 'medium',
      syncFrequency: json['syncFrequency'] ?? 'hourly',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'autoSync': autoSync,
      'syncOnWiFiOnly': syncOnWiFiOnly,
      'syncPhotos': syncPhotos,
      'syncPresets': syncPresets,
      'syncSettings': syncSettings,
      'compressionLevel': compressionLevel,
      'syncFrequency': syncFrequency,
    };
  }

  SyncSettings copyWith({
    bool? autoSync,
    bool? syncOnWiFiOnly,
    bool? syncPhotos,
    bool? syncPresets,
    bool? syncSettings,
    String? compressionLevel,
    String? syncFrequency,
  }) {
    return SyncSettings(
      autoSync: autoSync ?? this.autoSync,
      syncOnWiFiOnly: syncOnWiFiOnly ?? this.syncOnWiFiOnly,
      syncPhotos: syncPhotos ?? this.syncPhotos,
      syncPresets: syncPresets ?? this.syncPresets,
      syncSettings: syncSettings ?? this.syncSettings,
      compressionLevel: compressionLevel ?? this.compressionLevel,
      syncFrequency: syncFrequency ?? this.syncFrequency,
    );
  }
}

class CloudPhoto {
  final String id;
  final String filename;
  final String? originalFilename;
  final String s3Key;
  final String cloudUrl;
  final String? thumbnailUrl;
  final PhotoMetadata metadata;
  final String syncStatus;
  final double uploadProgress;
  final DateTime? capturedAt;
  final DateTime? uploadedAt;

  CloudPhoto({
    required this.id,
    required this.filename,
    this.originalFilename,
    required this.s3Key,
    required this.cloudUrl,
    this.thumbnailUrl,
    required this.metadata,
    required this.syncStatus,
    required this.uploadProgress,
    this.capturedAt,
    this.uploadedAt,
  });

  factory CloudPhoto.fromJson(Map<String, dynamic> json) {
    return CloudPhoto(
      id: json['_id'] ?? json['id'] ?? '',
      filename: json['filename'] ?? '',
      originalFilename: json['originalFilename'],
      s3Key: json['s3Key'] ?? '',
      cloudUrl: json['cloudUrl'] ?? '',
      thumbnailUrl: json['thumbnailUrl'],
      metadata: PhotoMetadata.fromJson(json['metadata'] ?? {}),
      syncStatus: json['syncStatus'] ?? 'pending',
      uploadProgress: (json['uploadProgress'] ?? 0.0).toDouble(),
      capturedAt: json['capturedAt'] != null ? DateTime.parse(json['capturedAt']) : null,
      uploadedAt: json['uploadedAt'] != null ? DateTime.parse(json['uploadedAt']) : null,
    );
  }

  bool get isSynced => syncStatus == 'synced';
  bool get isPending => syncStatus == 'pending';
  bool get isUploading => syncStatus == 'uploading';
  bool get hasFailed => syncStatus == 'failed';
}

class PhotoMetadata {
  final int? fileSize;
  final PhotoDimensions? dimensions;
  final String? format;
  final ExifData? exifData;
  final ProcessingInfo? processingInfo;

  PhotoMetadata({
    this.fileSize,
    this.dimensions,
    this.format,
    this.exifData,
    this.processingInfo,
  });

  factory PhotoMetadata.fromJson(Map<String, dynamic> json) {
    return PhotoMetadata(
      fileSize: json['fileSize'],
      dimensions: json['dimensions'] != null 
          ? PhotoDimensions.fromJson(json['dimensions'])
          : null,
      format: json['format'],
      exifData: json['exifData'] != null 
          ? ExifData.fromJson(json['exifData'])
          : null,
      processingInfo: json['processingInfo'] != null 
          ? ProcessingInfo.fromJson(json['processingInfo'])
          : null,
    );
  }
}

class PhotoDimensions {
  final int width;
  final int height;

  PhotoDimensions({required this.width, required this.height});

  factory PhotoDimensions.fromJson(Map<String, dynamic> json) {
    return PhotoDimensions(
      width: json['width'] ?? 0,
      height: json['height'] ?? 0,
    );
  }

  double get aspectRatio => height != 0 ? width / height : 1.0;
  String get displaySize => '${width}x$height';
}

class ExifData {
  final String? camera;
  final String? lens;
  final int? iso;
  final String? aperture;
  final String? shutterSpeed;
  final double? focalLength;
  final DateTime? capturedAt;
  final GpsLocation? gpsLocation;

  ExifData({
    this.camera,
    this.lens,
    this.iso,
    this.aperture,
    this.shutterSpeed,
    this.focalLength,
    this.capturedAt,
    this.gpsLocation,
  });

  factory ExifData.fromJson(Map<String, dynamic> json) {
    return ExifData(
      camera: json['camera'],
      lens: json['lens'],
      iso: json['iso'],
      aperture: json['aperture'],
      shutterSpeed: json['shutterSpeed'],
      focalLength: json['focalLength']?.toDouble(),
      capturedAt: json['capturedAt'] != null ? DateTime.parse(json['capturedAt']) : null,
      gpsLocation: json['gpsLocation'] != null 
          ? GpsLocation.fromJson(json['gpsLocation'])
          : null,
    );
  }
}

class GpsLocation {
  final double latitude;
  final double longitude;

  GpsLocation({required this.latitude, required this.longitude});

  factory GpsLocation.fromJson(Map<String, dynamic> json) {
    return GpsLocation(
      latitude: (json['latitude'] ?? 0.0).toDouble(),
      longitude: (json['longitude'] ?? 0.0).toDouble(),
    );
  }
}

class ProcessingInfo {
  final Map<String, dynamic>? aiAnalysis;
  final String? appliedPreset;
  final Map<String, dynamic>? adjustments;

  ProcessingInfo({
    this.aiAnalysis,
    this.appliedPreset,
    this.adjustments,
  });

  factory ProcessingInfo.fromJson(Map<String, dynamic> json) {
    return ProcessingInfo(
      aiAnalysis: json['aiAnalysis'],
      appliedPreset: json['appliedPreset'],
      adjustments: json['adjustments'],
    );
  }
}

// Extension for string capitalization
extension StringExtension on String {
  String capitalize() {
    if (isEmpty) return this;
    return "${this[0].toUpperCase()}${substring(1)}";
  }
}

// Helper for pow function